# Introduction
Welcome to Python Basics! Python is a popular high-level programming language known for its readability and versatility. It is widely used in data science, web development, automation, and many other fields. Python's syntax is beginner-friendly, making it a great language for newcomers to programming.

Topics covered include:
* Basic data types in Python (integers, floats, strings, booleans, lists, tuples, sets, dictionaries)
* Declaring and using variables; common operators (arithmetic, comparison, logical)
* Control flow structures (`if`, `elif`, `else` conditionals; `for` and `while` loops)
* Defining and calling functions (with parameters and return values)
* Simple real-world examples (temperature conversion, shopping cart price calculator)

*Note: This notebook is designed for use in a Jupyter/Colab environment. You can run the code cells below to see outputs and modify them to experiment. Remember to press `Shift+Enter` or click the `"RUN"` button to execute a code cell.*

# Data Types in Python
In programming, data types represent the kind of data you are working with. Python has several built-in data types to represent numbers, text, truth values, and collections of items. Let's explore some fundamental data types in Python:

## Numeric Types: Integers and Floats
**Integers** (`int`) are whole numbers (positive, negative, or zero) without a fractional part. For example, -3, 0, and 42 are integers. Python integers have unlimited precision, so you can work with very large integers as well.

**Floats** (`float`) are numbers with a decimal point (floating-point numbers), e.g. 3.14 or -0.001. They are used to represent real numbers (including fractions). Keep in mind that floating-point numbers are implemented in binary and can sometimes be an approximation of the true value.

Let's see integers and floats in action:

In [1]:
# Examples of integers and floats
a = 5          # int
b = 2.5        # float

print(a, "is of type", type(a))    # <class 'int'>
print(b, "is of type", type(b))    # <class 'float'>

# Arithmetic with int and float
sum_val = a + b
print("Sum:", sum_val)            # Sum: 7.5
diff = a - b
print("Difference:", diff)        # Difference: 2.5
product = a * b
print("Product:", product)        # Product: 12.5
quotient = a / b
print("Quotient:", quotient)      # Quotient: 2.0

5 is of type <class 'int'>
2.5 is of type <class 'float'>
Sum: 7.5
Difference: 2.5
Product: 12.5
Quotient: 2.0


In the code above, we created an integer `a` and a float `b`. We used the `type()` function to verify their types. We also performed basic arithmetic operations (addition, subtraction, multiplication, division) on these numbers. Notice that adding an `int` and a `float` results in a float (`5 + 2.5 = 7.5`).

Note: Python supports other numeric types as well, such as complex numbers (for complex arithmetic), but we'll focus on `int` and `float` here for simplicity.

## Strings (`str`)
Strings are sequences of characters used to represent text data. In Python, strings are enclosed in quotes, either single quotes `'...'` or double quotes `"..."`. For example, `"Hello, World!"` is a string. Strings can contain letters, numbers, symbols, and spaces.

You can concatenate (join) strings using the `+` operator, and repeat strings using the `*` operator. Strings are **immutable**, meaning that once created, their content cannot be changed (you can create new strings, but you cannot modify an existing string in place).

Let's see some string examples:

In [2]:
text1 = "Hello"
text2 = 'World'
combined = text1 + " " + text2   # Concatenation of strings
print(combined)                 # Hello World

excitement = "Woo!" * 3         # Repeat the string 3 times
print(excitement)               # Woo!Woo!Woo!

# Accessing characters in a string
name = "Python"
first_char = name[0]
print("First character:", first_char)        # First character: P
last_char = name[-1]
print("Last character:", last_char)          # Last character: n
print("Length of", name, "is", len(name))    # Length of Python is 6


Hello World
Woo!Woo!Woo!
First character: P
Last character: n
Length of Python is 6


In this example, we concatenated two strings with a space in between to form `"Hello World"`. We also repeated a string multiple times. We then accessed characters by index: `name[0]` gives the first character `'P'`, and `name[-1]` gives the last character `'n'`. We used the `len()` function to get the length of the string `"Python"`, which is 6 characters.


## Booleans (`bool`)
Booleans represent truth values and can only be either `True` or `False`. Booleans often result from **comparison operations** or logical conditions. In Python, `True` and `False` are reserved keywords.

Booleans are very useful in controlling program flow (as we'll see with conditionals and loops). They are actually a subtype of integers in Python (internally `True` is `1` and `False` is `0`), but they behave like distinct logical values.

Examples of booleans:

In [3]:
is_raining = False
is_sunny = True
print("is_raining is", is_raining)   # is_raining is False
print("is_sunny is", is_sunny)       # is_sunny is True

# Booleans from comparisons:
print(5 > 3)   # True, because 5 is greater than 3
print(5 < 3)   # False, because 5 is not less than 3
print(7 == 7)  # True, 7 is equal to 7
print(7 != 7)  # False, 7 is not different from 7

is_raining is False
is_sunny is True
True
False
True
False


In this code, we set two boolean variables and printed them. We also demonstrated how comparison operators (like `>`, `<`, `==`, `!=`) produce boolean results. For instance, `5 > 3` is `True` and `5 < 3` is `False`. We will talk more about these operators in the Operators section.

## Lists (`list`)
Lists are ordered collections of items (elements) enclosed in square brackets `[]`. Lists can contain elements of any type, and the elements can be accessed by their index (position in the list, starting from 0). Lists are **mutable**, meaning you can change, add, or remove elements after the list is created.

Here are some examples of using lists:

In [4]:
# Create a list of numbers
numbers = [10, 20, 30, 40, 50]
print("Numbers:", numbers)           # Numbers: [10, 20, 30, 40, 50]
print("First element:", numbers[0])  # First element: 10
print("Last element:", numbers[-1])  # Last element: 50

# Lists can contain mixed data types
random_list = [1, "two", 3.0, True]
print("Mixed list:", random_list)    # Mixed list: [1, 'two', 3.0, True]

# Modify elements in a list
numbers[2] = 99
print("Modified numbers:", numbers)  # Modified numbers: [10, 20, 99, 40, 50]

# Add elements to a list
numbers.append(60)
print("After appending:", numbers)   # After appending: [10, 20, 99, 40, 50, 60]

# Slice a list (get a sublist)
sub_list = numbers[1:4]  # elements from index 1 to 3 (excluding 4)
print("Slice of list:", sub_list)    # Slice of list: [20, 99, 40]

Numbers: [10, 20, 30, 40, 50]
First element: 10
Last element: 50
Mixed list: [1, 'two', 3.0, True]
Modified numbers: [10, 20, 99, 40, 50]
After appending: [10, 20, 99, 40, 50, 60]
Slice of list: [20, 99, 40]


We created a list of integers and accessed elements by index. We then showed that lists can have mixed types (though typically you'll use the same type for all elements in a list). We modified an element at index 2, changing `30` to `99`. We used `append()` to add a new element to the end of the list. Finally, we demonstrated list slicing: `numbers[1:4]` gives a new list with elements from index 1 up to (but not including) index 4.

## Tuples (`tuple`)
**Tuples** are very similar to lists in that they are ordered collections of elements. The key difference is that tuples are **immutable** – once a tuple is created, its elements cannot be changed. Tuples are defined using parentheses `()` (or simply by separating values with commas). Tuples are often used for fixed collections of items that should not change throughout the program. For example, a pair of coordinates (x, y) can be stored in a tuple.

Example of a tuple:

In [5]:
coordinates = (12.5, 7.8)
print("Coordinates:", coordinates)       # Coordinates: (12.5, 7.8)
print("X coordinate:", coordinates[0])   # X coordinate: 12.5
print("Y coordinate:", coordinates[1])   # Y coordinate: 7.8

# Tuples can also be created without parentheses by just using commas
values = 1, 2, 3
print("Tuple values:", values)           # Tuple values: (1, 2, 3)

# Attempting to modify a tuple (will cause an error if uncommented)
immutable_tuple = (5, 6, 7)
# immutable_tuple[0] = 10  # Uncommenting this line would result in a TypeError

Coordinates: (12.5, 7.8)
X coordinate: 12.5
Y coordinate: 7.8
Tuple values: (1, 2, 3)


We created a tuple `coordinates` and accessed its elements. We also showed that you can define a tuple by listing values separated by commas (parentheses are optional in that case, but often included for clarity). If you try to modify a tuple (as shown in the commented line), Python will raise a `TypeError` because tuples cannot be changed after creation.

## Sets (`set`)
Sets are unordered collections of unique elements. They are defined using curly braces `{}` or the `set()` constructor. A set automatically eliminates duplicate values and is useful when you want to store items without any particular order and ensure no duplicates. Sets are mutable (you can add or remove elements), but since they are unordered, they do not support indexing like lists or tuples.

Example of using a set:

In [6]:
# Define a set of fruits
fruits = {"apple", "banana", "cherry", "apple"}
print("Fruits set:", fruits)    # Fruits set: {'banana', 'apple', 'cherry'}

# Note: 'apple' was included twice, but appears only once in the set (no duplicates).

# Add and remove elements in a set
fruits.add("orange")
print("After adding orange:", fruits)   # After adding orange: {'banana', 'apple', 'cherry', 'orange'}
fruits.remove("banana")
print("After removing banana:", fruits) # After removing banana: {'apple', 'cherry', 'orange'}

# Check membership in a set
print("Is 'apple' in fruits?", "apple" in fruits)   # Is 'apple' in fruits? True
print("Is 'banana' in fruits?", "banana" in fruits) # Is 'banana' in fruits? False

Fruits set: {'cherry', 'apple', 'banana'}
After adding orange: {'orange', 'cherry', 'apple', 'banana'}
After removing banana: {'orange', 'cherry', 'apple'}
Is 'apple' in fruits? True
Is 'banana' in fruits? False


We created a set of fruit names. When printing the set, notice that the order may not be the same as the order we added elements (sets are unordered). Also, even though we added `"apple"` twice, it only shows up once in the set output, because sets cannot have duplicates. We demonstrated adding an element with `add()`, removing an element with `remove()`, and checking membership with the `in` keyword.

## Dictionaries (`dict`)
Dictionaries are collections of key-value pairs. Each item in a dictionary has a key and a value. You can use the key to retrieve the corresponding value. Dictionaries are created using curly braces `{key: value, ...}` or the `dict()` constructor, and they are mutable.

In a dictionary, keys are typically strings or numbers, and values can be of any type. Keys must be unique and immutable (for example, you can use numbers, strings, or tuples as keys, but not lists or other dictionaries).

Example of a dictionary:

In [7]:
# Create a dictionary of student ages
student_ages = {"Alice": 25, "Bob": 22, "Charlie": 23}
print("Student ages:", student_ages)
print("Alice's age:", student_ages["Alice"])   # Alice's age: 25

# Add a new key-value pair
student_ages["David"] = 24
print("After adding David:", student_ages)

# Modify an existing value
student_ages["Alice"] = 26
print("After updating Alice's age:", student_ages)

# Remove a key-value pair
del student_ages["Bob"]
print("After removing Bob:", student_ages)

Student ages: {'Alice': 25, 'Bob': 22, 'Charlie': 23}
Alice's age: 25
After adding David: {'Alice': 25, 'Bob': 22, 'Charlie': 23, 'David': 24}
After updating Alice's age: {'Alice': 26, 'Bob': 22, 'Charlie': 23, 'David': 24}
After removing Bob: {'Alice': 26, 'Charlie': 23, 'David': 24}


In this example, we created a dictionary where the keys are student names and the values are their ages. We accessed the age of "Alice" using the key `"Alice"`. We then added a new entry for "David", updated Alice's age, and removed Bob's entry using the `del` statement. After each operation, we printed the dictionary to see the changes.

**Note**: In Python 3.7 and above, dictionaries maintain the insertion order of keys. However, you should primarily think of dictionaries as unordered collections for older versions of Python or conceptual purposes.

## Practice Exercises: Data Types
Try these exercises to practice what you've learned about data types:
1. **List Practice:** Create a list of five different numbers. Then, do the following:
Print the list.
Access and print the third element (index `2`) of the list.
Change the last element of the list to a new value, and print the modified list.
2. **Dictionary Practice:** Create a dictionary to store the names and ages of three people. For example, `{"Alice": 30, "Bob": 25, "Eve": 29}`. Then:
  * Print the dictionary.
  * Retrieve and print the age of one person from the dictionary.
  * Add a new person to the dictionary, and print the updated dictionary.
3. **Set Practice:** Create a set of several words, with at least one repeated word. Print the set and observe the result. Then:
  * Add a new word to the set.
  * Remove an existing word from the set.
  * Check whether a certain word is `in` the set (using the in operator) and print the result.
  
Feel free to create new code cells below and try these exercises out.

## Practice Exercises: Data Types
Try these exercises to practice what you've learned about data types:
1. **List Practice:** Create a list of five different numbers. Then, do the following:
Print the list.
Access and print the third element (index `2`) of the list.
Change the last element of the list to a new value, and print the modified list.
2. **Dictionary Practice:** Create a dictionary to store the names and ages of three people. For example, `{"Alice": 30, "Bob": 25, "Eve": 29}`. Then:
    * Print the dictionary.
    * Retrieve and print the age of one person from the dictionary.
    * Add a new person to the dictionary, and print the updated dictionary.
3. **Set Practice:** Create a set of several words, with at least one repeated word. Print the set and observe the result. Then:
    * Add a new word to the set.
    * Remove an existing word from the set.
    * Check whether a certain word is in the set (using the `in` operator) and print the result.

Feel free to create new code cells below and try these exercises out.

### Variables and Operators
In Python, a **variable** is a name that refers to a value. You can imagine a variable as a labeled box that holds some data. Unlike some languages, Python does not require you to declare the type of a variable upfront – the type is inferred from the value you assign to it. You can even reassign a variable to a value of a different type later (though it's good practice to keep the type consistent for a given variable).
#### Using Variables
To create a variable, simply assign a value to a name using the equals sign `=` (this is called the assignment operator). Variable names can consist of letters, numbers, and underscores, but cannot start with a number. They are case-sensitive (e.g., `Data` and `data` would be different variables).

 Example of declaring and using variables:

In [8]:
x = 10            # assign 10 to x
y = 3.5           # assign 3.5 to y
name = "Alice"    # assign "Alice" to name

print(x)          # 10
print(y)          # 3.5
print(name)       # Alice

# Variables can be reassigned to new values (even of a different type)
x = "Now I'm a string!"
print(x)          # Now I'm a string!

10
3.5
Alice
Now I'm a string!


In this example, we created an integer variable `x`, a float variable `y`, and a string variable `name`. We printed their values. Then we reassigned `x` to a string value. Python allowed this because it is **dynamically-typed** (the type of `x` changed from int to str when we assigned a string to it). However, changing types like that can be confusing, so it's generally better to use variables consistently.

**Tip:** It's good practice to choose descriptive variable names so that your code is easier to understand (e.g., use `age = 30` instead of `a = 30`).

## Arithmetic Operators
Python supports the typical arithmetic operators for numerical calculations:
  * `+` (addition)
  * `-` (subtraction)
  * `*` (multiplication)
  * `/` (division)
  * `//` (integer/floor division)
  * `%` (modulus, i.e., remainder of division)
  * `**` (exponentiation, e.g., power calculation)

Let's use these operators in examples:

In [10]:
a = 15
b = 4
print("a + b =", a + b)   # a + b = 19
print("a - b =", a - b)   # a - b = 11
print("a * b =", a * b)   # a * b = 60
print("a / b =", a / b)   # a / b = 3.75 (normal division)
print("a // b =", a // b) # a // b = 3   (floor division, quotient of 15/4)
print("a % b =", a % b)   # a % b = 3    (remainder of 15/4)
print("a ** b =", a ** b) # a ** b = 50625 (15 to the power of 4)

a + b = 19
a - b = 11
a * b = 60
a / b = 3.75
a // b = 3
a % b = 3
a ** b = 50625


Important notes:
* **Division (`/`)** always produces a float result (even if the division is even). For example, `8 / 2` will give `4.0`, not `4`.
* **Floor division (`//`)** gives the integer quotient without the remainder (it "floors" the result). For example, `15 // 4` is `3`.
* **Modulus (`%`)** gives the remainder after division. For example, `15 % 4` is `3` (because 4 * 3 = 12, and the remainder to get to 15 is 3).
Exponentiation (`**`) is used for powers (e.g., `2 ** 3` is `8`).

##Comparison Operators
Comparison operators are used to compare two values and result in a boolean (True or False):
* `==` : equals (checks if two values are equal)
* `!=` : not equals (checks if two values are different)
* `>` : greater than
* `<` : less than
* `>=` : greater than or equal to
* `<=` : less than or equal to


Examples of comparison operations:

In [11]:
x = 10
y = 7
print("x == y ?", x == y)   # False, 10 is not equal to 7
print("x != y ?", x != y)   # True, 10 is not equal to 7
print("x > y  ?", x > y)    # True, 10 is greater than 7
print("x < y  ?", x < y)    # False, 10 is not less than 7
print("x >= 10?", x >= 10)  # True, 10 is greater or equal to 10
print("y <= 7 ?", y <= 7)   # True, 7 is less or equal to 7

x == y ? False
x != y ? True
x > y  ? True
x < y  ? False
x >= 10? True
y <= 7 ? True


Each of these operations produces a boolean result. For instance, since `x` is 10 and `y` is 7, `x > y` is `True` but `x < y` is `False`. We also see that `x >= 10` is `True` (because x is exactly 10), and `y <= 7` is `True` (because y is exactly 7).

## Logical Operators
Logical operators are used to combine boolean values (or expressions) and also result in a boolean. Python has three logical operators:
* `and` : True if **both** operands are True
* `or` : True if **at least one** operand is True
* `not` : True if the operand is False (and vice versa; it inverts the boolean value)

These are often used to build complex conditions. For example, checking multiple conditions in an `if` statement.

Examples of logical operations:

In [12]:
condition1 = True
condition2 = False
print("condition1 and condition2:", condition1 and condition2)  # False (because condition2 is False)
print("condition1 or condition2:", condition1 or condition2)    # True (because at least one is True)
print("not condition1:", not condition1)                        # False (condition1 is True, so not True is False)

# Using logical operators with comparison expressions:
x = 5
print(x > 0 and x < 10)   # True (x is between 0 and 10)
print(x < 0 or x == 5)    # True (one of the conditions is True: x == 5)

condition1 and condition2: False
condition1 or condition2: True
not condition1: False
True
True


In this code, we first demonstrated `and`, `or`, and `not` with simple True/False values. Then we used logical operators in combination with comparison operators: for example, `x > 0` and `x < 10` is True only if both conditions are True (which in this case they are, since `x` is 5).

### Practice Exercises: Variables and Operators
Test your understanding of variables and operators with these exercises:
1. **Simple Calculator:** Create two variables `a` and `b`, assign them any numbers you like, and then print out the results of `a + b`, `a - b`, `a * b`, and `a / b`.
2. **Rectangle Area:** Suppose you have a rectangle of width 5 and height 3 (store these in variables). Calculate the area of the rectangle (width * height) and print the result.
3. **Even or Odd?:** Write a script that defines a variable `num` and assigns an integer to it. Then use the modulus operator `(%)` to check if the number is even or odd. Print `"even"` if it's even or `"odd"` if it's odd.
4. **Range Check:** Define a variable `x` and assign a number to it. Write an expression using comparison and logical operators to check if `x` is between 1 and 100 (inclusive). The expression should print `True` if it is, and `False` otherwise.

Try to write and run code for each of these in new cells.

## Control Flow
Control flow structures allow you to direct the order of execution of statements in your program based on conditions and loops. Python uses indentation (whitespace at the beginning of a line) to define code blocks (such as the body of an `if` or the contents of a loop). This is different from some other languages that use braces `{}` to delineate blocks.

Let's explore conditional statements and loops in Python.

### Conditional Statements: `if`, `elif`, `else`
Conditional statements allow you to execute certain pieces of code only if specific conditions are met. In Python:
  * Use `if` to specify a block of code to run if a condition is True.
  * Optionally use `elif` (short for "else if") to specify additional conditions to check if the previous ones were False.
  * Optionally use `else` to specify a block of code to run if all previous conditions are False.

**Important:** The condition expressions must be followed by a colon `:`, and the code block under each `if/elif/else` must be indented (typically by 4 spaces or a tab). Consistent indentation is crucial—Python will give an error if your indentation is inconsistent or incorrect.

Example of using if-elif-else:

In [13]:
temperature = 30  # degrees Celsius
if temperature > 30:
    print("It's hot outside.")
elif temperature < 10:
    print("It's cold outside.")
else:
    print("The temperature is moderate.")

The temperature is moderate.


In this code, if `temperature` is greater than 30, the first `print` will execute. If not, but `temperature` is less than 10, the `elif` branch executes. If neither condition is met (meaning 10 <= temperature <= 30), the `else` branch executes. Try changing the value of temperature and running the code to see how the output changes.

Here's another example demonstrating an if-else (with no elif) to check if a number is even or odd:

In [14]:
num = 42
if num % 2 == 0:
    print(num, "is even")
else:
    print(num, "is odd")

42 is even


The condition `num % 2 == 0` checks if the number has no remainder when divided by 2 (which means it's even). Depending on the result, it prints the appropriate message.

### `for` Loops
A **for loop** is used to iterate over a sequence (such as a list, tuple, string, or range of numbers) and execute a block of code for each element in the sequence. In Python, a `for` loop takes an item from the sequence in each iteration and assigns it to a loop variable that you specify.

Commonly, we use `for` loops with the `range()` function when we want to loop a specific number of times. `range(n)` produces a sequence of numbers from 0 up to n-1.

Examples of for loops:

In [15]:
# Example 1: Iterating over a list
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print("I love", fruit)

I love apple
I love banana
I love cherry


This loop will print each fruit in the list with the phrase "I love" in front of it. The variable `fruit` takes each value from the `fruits` list in turn.

In [16]:
# Example 2: Using range() in a for loop
for i in range(5):
    print("Index:", i)

Index: 0
Index: 1
Index: 2
Index: 3
Index: 4


In this loop, `range(5)` generates the sequence `0, 1, 2, 3, 4`. The loop variable `i` will take each of these values, and the loop will run 5 times, printing the current index each time.

In [17]:
# Example 3: Summing numbers with a for loop
total = 0
for num in [1, 2, 3, 4, 5]:
    total += num  # add each number to total
print("Sum of 1-5:", total)  # Sum of 1-5: 15

Sum of 1-5: 15


Here we initialized `total` to 0, then looped through a list of numbers 1 through 5, adding each number to `total`. After the loop, `total` contains the sum of those numbers.

### `while` Loops
A while loop repeats a block of code as long as a condition remains True. It has the form:

In [20]:
# while condition:
#     # code block

The loop will check the condition at the start of each iteration, and if the condition is True, it executes the code block, then repeats. If the condition becomes False, it exits the loop.

Be careful to ensure that the condition will eventually become False; otherwise, you'll get an infinite loop (which will make your program run forever until you stop it).

Example of a while loop:

In [19]:
count = 1
while count <= 5:
    print("Count is", count)
    count += 1  # increment count by 1

Count is 1
Count is 2
Count is 3
Count is 4
Count is 5


This loop will print the count from 1 to 5. Here's what happens:
* Initially, `count` is 1. The condition `count <= 5` is True, so it enters the loop.
* It prints "Count is 1", then increments `count` to 2.
* The loop condition is checked again; now `count` is 2 (still <= 5), so the loop continues.
* This repeats until `count` becomes 6. At that point, `count <= 5` is False, so the loop stops.

**Tip:** If you find yourself writing a loop that never ends (infinite loop), use the stop button in your notebook or press `Ctrl+C` in a terminal to interrupt execution.

### Practice Exercises: Control Flow
Use what you've learned about conditionals and loops to solve these problems:
1. **Grade Checker:** Write an `if-elif-else` structure that assigns a letter grade based on a numeric score. For example, if `score >= 90` print "A", `score >= 80` print "B", `score >= 70` print "C", `score >= 60` print "D", else print "F". Try changing the score value to test different branches.
2.**Even Numbers with a For Loop:** Use a `for` loop with `range()` to print all even numbers from 2 up to 20 (inclusive).
3 . **Countdown with While:** Use a `while` loop to print a countdown from 5 to 1, then print "Blast off!" after the loop completes.
4. **Sum with While:** Write a while loop to calculate the sum of numbers from 1 to 100. (Hint: Initialize a counter and a total, and increment the counter each loop iteration.)

Try writing and executing code for each of these exercises.

### Functions
Functions are reusable blocks of code that perform a specific task. You can **define** a function once and then **call it** (execute it) multiple times with different inputs. Functions help organize code, make it more readable, and avoid repetition.

In Python, we define a function using the `def` keyword, followed by the function name and parentheses `()`. Inside the parentheses, we can define parameters (variables that will receive the data passed into the function). The function body is indented and typically contains a return statement to return a result (though functions can also just perform actions without returning anything, in which case they implicitly return `None`).

The basic syntax is:

In [22]:
# def function_name(parameter1, parameter2, ...):
#     """
#     Optional docstring (description of function).
#     """
#     # function body
#     return value

Now let's see some examples of functions:

In [23]:
# Example 1: A simple function with no parameters that prints a message
def greet():
    print("Hello, welcome!")

greet()  # Calling the function

Hello, welcome!


In this example, we defined a function named `greet` that takes no parameters and simply prints a greeting. When we call `greet()`, it executes the print statement.

In [24]:
# Example 2: Function with one parameter
def greet_person(name):
    # name is a parameter that the function will use
    print("Hello,", name + "!")

greet_person("Alice")  # Hello, Alice!
greet_person("Bob")    # Hello, Bob!

Hello, Alice!
Hello, Bob!


Here, `greet_person` is a function that takes one parameter `name`. It uses that parameter inside the function to print a personalized greeting. We called the function twice with different arguments ("Alice" and "Bob").

In [26]:
# Example 3: Function with a return value
def add(a, b):
    # This function takes two numbers and returns their sum
    result = a + b
    return result

sum_value = add(5, 7)
print("Sum of 5 and 7 is", sum_value)  # Sum of 5 and 7 is 12
print("Sum of 10 and 20 is", add(10, 20))  # Sum of 10 and 20 is 30

Sum of 5 and 7 is 12
Sum of 10 and 20 is 30


In `add(a, b)`, we take two parameters, add them, and use `return` to send back the result. We can then store that result in a variable (like `sum_value`) or use it directly in an expression or print statement. The second print demonstrates calling `add(10, 20)` directly inside the print function.

### Why use functions?
* **Reusability**: Write once, use multiple times. If you need to perform a task in multiple places, putting it in a function avoids repeating code.
* **Organization**: Functions help break a program into smaller, logical pieces. Each function can do a clear, specific thing, making the code easier to understand.
* **Maintainability**: If something needs to change (e.g., a formula), you only have to update it in the function rather than in many places in your code.


Now, let's use functions in some real-world style examples.

## Real-World Examples
The following examples demonstrate how you can apply the basics you've learned to simple real-world problems.
### Example: Temperature Conversion
Suppose you want to convert temperatures from Celsius to Fahrenheit. The formula for conversion is:
Fahrenheit=9/5 × Celsius +32

We can write a function to do this conversion for us.

In [27]:
def celsius_to_fahrenheit(celsius):
    """Convert Celsius to Fahrenheit."""
    return (celsius * 9/5) + 32

# Test the function with a few values
print("0°C in Fahrenheit is:", celsius_to_fahrenheit(0))     # 32.0
print("37°C in Fahrenheit is:", celsius_to_fahrenheit(37))   # 98.6
print("100°C in Fahrenheit is:", celsius_to_fahrenheit(100)) # 212.0

0°C in Fahrenheit is: 32.0
37°C in Fahrenheit is: 98.6
100°C in Fahrenheit is: 212.0


We defined a function `celsius_to_fahrenheit` that takes a Celsius temperature and returns the equivalent Fahrenheit temperature. We then tested the function with a few values: 0°C (freezing point of water, which is 32°F), 37°C (about human body temperature, ~98.6°F), and 100°C (boiling point of water, 212°F). Using functions like this makes our code cleaner and more modular. If we need to convert many temperatures, we can just call this function for each value instead of writing the formula repeatedly.

### Example: Shopping Cart Price Calculator
Imagine an online shopping scenario where you have a list of items in your cart and you want to calculate the total price. We can use a dictionary to store the prices of each item, and a list to represent the shopping cart (as a list of items purchased). Then, we can use a loop to calculate the total cost.

In [28]:
# Price list for items (dictionary of item: price)
prices = {
    "apple": 0.5,
    "banana": 0.25,
    "milk": 2.5,
    "bread": 1.5
}

# Shopping cart (list of items we want to buy)
cart = ["apple", "apple", "bread", "milk"]

# Calculate total cost
total_cost = 0
for item in cart:
    # Add the price of each item in the cart to total_cost
    total_cost += prices[item]
print("Cart items:", cart)
print("Total cost: $" + str(total_cost))

Cart items: ['apple', 'apple', 'bread', 'milk']
Total cost: $5.0


In this code:
* We defined a dictionary `prices` that holds the price of each item.
* We defined a list `cart` with some items (in this case, 2 apples, 1 bread, and 1 milk).
* We then looped through each item in the cart, looked up its `price` in the prices dictionary, and added that to `total_cost`.
* Finally, we printed out the items in the cart and the total cost.

If you run this code, you should see that the total cost for `["apple", "apple", "bread", "milk"]` is the sum of the prices: two apples ($0.5 each) + one bread ($1.5) + one milk ($2.5) = $0.5 + $0.5 + $1.5 + $2.5 = $5.0.

This simple example shows how dictionaries and lists can work together in a practical task. In a real application, you might generate the cart list from user input or a database, and you might have many more items and more complex logic (tax, discounts, etc.), but the basic idea of summing up prices remains the same.

### Practice Exercises: Functions
Now that you've seen how to define and use functions, try writing your own:

1. **Circle Area Function:** Write a function `calculate_area(radius)` that takes the radius of a circle and returns the area of the circle. (Assume π = 3.14159 or use the `math` module's `math.pi` value for more precision.) Then, call your function with different radii to test it (e.g., radius 1, 5, 10).
2. **Maximum of Three:** Write a function `max_of_three(a, b, c)` that returns the largest of three numbers. Try calling it with various inputs to ensure it works for all cases.
3. **Odd or Even Function:** Write a function `is_even(number)` that returns `True` if the given number is even and `False` if it is odd. (Hint: use the modulus operator %). Test your function with both even and odd numbers.
4. **List Sum Function:** (Challenge) Write a function `sum_list(numbers)` that takes a list of numbers and returns the sum of those numbers. Try it with a few examples, and perhaps compare with Python's built-in `sum()` function for verification.

## **Conclusion**
In this notebook, we covered the basics of Python programming, including data types, variables, operators, control flow, and functions. These fundamentals are the building blocks of writing any Python program. By practicing the concepts and exercises provided, you should now be able to write simple scripts and functions on your own. This is just the beginning of your Python journey. With a solid grasp of the basics, you're ready to explore more advanced topics like working with libraries for data science (NumPy, pandas, etc.), data visualization, and more. Keep experimenting and happy coding!